src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +383,299 @@ public static void EnsureSocketsAreInitialized()
                             throw new SocketException((int)errorCode);
                         }
 
+                        s_getAddrInfoExSupported = TestGetAddrInfoEx();
+
                         Volatile.Write(ref s_initialized, true);
                     }
                 }
             }
         }
+
+        private static bool TestGetAddrInfoEx()
+        {
+            using (SafeLibraryHandle libHandle = Interop.Kernel32.LoadLibraryExW(Interop.Libraries.Ws2_32, IntPtr.Zero, 0))
+            {
+                return Interop.Kernel32.GetProcAddress(libHandle, nameof(Interop.Winsock.GetAddrInfoExCancel)) != IntPtr.Zero;
+            }
+        }
+
+        public static unsafe void GetAddrInfoAsync(string name, DnsResolveAsyncResult asyncResult)
+        {
+            try
+            { }
+            finally
+            {
+                GetAddrInfoExContext* context = GetAddrInfoExContext.AllocateContext(name, asyncResult);
+
+                AddressInfoEx hints = new AddressInfoEx();
+                hints.ai_flags = AddressInfoHints.AI_CANONNAME;
+                hints.ai_family = AddressFamily.Unspecified; // Gets all address families
+
+                SocketError errorCode =
+                    (SocketError)Interop.Winsock.GetAddrInfoExW(name, null, 0 /* NS_ALL*/, IntPtr.Zero, ref hints, out context->Result, IntPtr.Zero, ref context->Overlapped, s_getAddrInfoExCallback, out context->CancelHandle);
+
+                if (errorCode != SocketError.IOPending)
+                    ProcessResult(errorCode, context);
+            }
+        }
+
+        private static unsafe void GetAddressInfoExCallback([In] int error, [In] int bytes, [In] NativeOverlapped* overlapped)
+        {
+            try
+            { }
+            finally
+            {
+                // Can be casted directly to QueryContext* because the overlapped is its first field
+                GetAddrInfoExContext* context = (GetAddrInfoExContext*)overlapped;
+
+                ProcessResult((SocketError)error, context);
+            }
+        }
+
+        private static unsafe void ProcessResult(SocketError errorCode, GetAddrInfoExContext* context)
+        {
+            GetAddrInfoExState state = context->GetQueryState();
+
+            if (state == null || !state.SetCallbackStartedOrCanceled())
+                return;
```

Comment:
Who cleans up the allocated context?

Reply:
The context is freed at the end of ProcessResult or in GetAddrInfoExState's finalizer when the AppDomain is unloaded.

The callback should not be called twice, but the callback could be raised at the same time the AppDomain is unloaded, so this check ensure only one path will free the context.

I will add a comment.